### PR TITLE
[QOL] Quick Upgrade Utility Implementation

### PR DESCRIPTION
**Pull request by zethnest**
ORIGINAL:
Peek 2020-05-28 12-03
UPDATE 1:
Peek 2020-05-30 10-30
UPDATE 2: Mobile
Peek 2020-05-31 18-02

A preview of what it looks like.

How it works using the default buttons, you hold down `Left Mouse Button`, and then hold down `Middle Mouse Button`, and then you drag across. If you want to force `Thorium Tier` blocks like ArmoredConveyor and ArmoredConduit to every block, you press `Left Control`.

As for mobile. When hammer is toggled, toggling it again will toggle the upgrade. You can select as shown in the desktop version, long press, then drag across. As for forced mode in desktop version, you can toggle the lock icon at the bottom left.

~~Currently doesn't work with multiblocks as shown in the end of the GIF.~~

An alternative implementation to https://github.com/Anuken/Mindustry/pull/1009 and https://github.com/Anuken/Mindustry/pull/1722

EDIT: Some questions from Discord if I ever do a mobile version. Since this is a draft, I can put it on the TODO list. Originally I wanted to implement for desktop version first, mobile version in another PR. 

TODO:
- [X] Support Multiblocks
- [X] Mobile Version Support

### CODE REVIEW

File: core/src/mindustry/input/InputHandler.java [hunk 1]
```diff
@@ -601,6 +617,44 @@ protected void removeSelection(int x1, int y1, int x2, int y2, boolean flush){
...
+            for(int y = dresult.y; y <= dresult.y2; y++){
+                Tile tile = world.tilec(x, y);
+                if(tile.block() == Blocks.air || tile.block() instanceof Rock
+                    || !validPlace(x, y, block, tile.rotation())
+                    || !(tile.block() instanceof Conveyor || tile.block() instanceof Conduit || tile.block() instanceof Wall)) continue;
+
+                if((block instanceof ArmoredConveyor || block instanceof ArmoredConduit) && !forced){
+                    if(tile.left() != null){
+                        Tilec left = tile.left();
+                        if(((Autotiler)tile.block()).blends(tile, tile.rotation(), left.tileX(), left.tileY(), left.rotation(), left.block())) continue;
```
Comment (zethnest) [inline core/src/mindustry/input/InputHandler.java:634]: Somehow worked the first time, then doesn't after though it should after research. Going to remove in next commit

File: core/src/mindustry/input/InputHandler.java [hunk 2]
```diff
@@ -601,6 +617,44 @@ protected void removeSelection(int x1, int y1, int x2, int y2, boolean flush){
...
+        lineRequests.clear();
+
+        NormalizeResult dresult = Placement.normalizeArea(x1, y1, x2, y2, rotation, false, maxLength);
+        for(int x = dresult.x; x <= dresult.x2; x++){
+            for(int y = dresult.y; y <= dresult.y2; y++){
+                Tile tile = world.tilec(x, y);
+                if(tile.block() == Blocks.air || tile.block() instanceof Rock
+                    || !validPlace(x, y, block, tile.rotation())
+                    || !(tile.block() instanceof Conveyor || tile.block() instanceof Conduit || tile.block() instanceof Wall)) continue;
+
```
Comment (Quezler) [inline core/src/mindustry/input/InputHandler.java:630]: ima just leave this comment from anuke here: https://github.com/Anuken/Mindustry/pull/950#discussion_r341330988

Comment (zethnest) [reply core/src/mindustry/input/InputHandler.java:630]: I'll try to use it in this implementation. Thank you for guidance.

File: core/src/mindustry/world/blocks/liquid/Conduit.java [hunk 3]
```diff
@@ -57,6 +57,12 @@ public Block getReplacement(BuildRequest req, Array<BuildRequest> requests){
             Mathf.mod(req.tile().rotation() - req.rotation, 2) == 1 ? Blocks.liquidJunction : this;
     }
 
+    @Override
+    public Block upgrade(Tile tile){
+        return tile.block() != null && tile.block() instanceof Conduit ? this : null;
+    }
+
+
```
Comment (Quezler) [inline core/src/mindustry/world/blocks/liquid/Conduit.java:65]:
```suggestion

```

File: core/src/mindustry/input/InputHandler.java [hunk 4]
```diff
@@ -601,6 +613,32 @@ protected void removeSelection(int x1, int y1, int x2, int y2, boolean flush){
...
+                if(upgrade == null || !validPlace(x, y, upgrade, tile.rotation())) continue;
+
+                BuildRequest req = new BuildRequest(x, y, tile.rotation(), upgrade);
+                req.animScale = 1f;
+                lineRequests.add(req);
+            }
+        }
+    }
+
+
```
Comment (Quezler) [inline core/src/mindustry/input/InputHandler.java:633]:
```suggestion

```

File: core/src/mindustry/input/DesktopInput.java [hunk 5]
```diff
@@ -449,8 +462,13 @@ void pollInput(){
             updateLine(selectX, selectY);
             lastLineX = cursorX;
             lastLineY = cursorY;
+        }else if((cursorX != lastLineX || cursorY != lastLineY) && isPlacing() && mode == upgrading){
+            updateArea(selectX, selectY);
+            lastLineX = cursorX;
+            lastLineY = cursorY;
         }
 
+
```
Comment (Quezler) [inline core/src/mindustry/input/DesktopInput.java:471]:
```suggestion
```

**Comment by garyType1** [comment]
Hmm, great idea. Though what's with the lock icon

**Comment by zethnest** [comment]
Should represent forced or not. With it is locked, armored conveyor and
conduit will ignore junctions. Unlocked will respect junctions.

**Comment by Anuken** [comment]
This PR is just too extensive and introduces too much code for a feature that isn't going to be used very often. In many cases, you can just drag the block over its predecessor in a straight line - making an area selection for upgrades is overkill.

**Pull request closed without merging**